### PR TITLE
prov/efa: refactor send of multiple packets of same packet type

### PR DESCRIPTION
**Pull request by wzamazon**
For certain packet types such as MEDIUM_RTM, RUNTREAD_RTM and DATA, we send out multiple packets with one call to `ibv_post_send`. Current code use the FI_MORE flag to control the behavior, which require the ep to maintain an internal list of packets that are pending send. In current setting, 1 call to `rxr_pkt_entry_send` does not necessary lead to 1 call to `ibv_post_send`.

This PR has multiple patches that eventually that eliminate the internal list, thus one call to `rxr_pkt_entry_send` will result in 1 call to `ibv_post_send`.

### CODE REVIEW

**Comment by wenduwan** [comment]
> This PR has multiple patches that eventually that eliminate the internal list, thus one call to rxr_pkt_entry_send will result in 1 call to ibv_post_send.

Is there any limitation to `ibv_post_send` that would require us to redrive the send? What is the benefit of fewer `ibv_post_send` calls?

**Comment by wzamazon** [comment]
>Is there any limitation to ibv_post_send that would require us to redrive the send? 

Not sure I understand the question.

> What is the benefit of fewer ibv_post_send calls?

1 call to `ibv_post_send` with multiple buffers would allow the hardware to optimize send more.

**Comment by wenduwan** [comment]
>    Is there any limitation to ibv_post_send that would require us to redrive the send?

I'm curious if `ibv_post_send` is able to handle arbitrarily long packet lists.

**Comment by wzamazon** [comment]
> I'm curious if ibv_post_send is able to handle arbitrarily long packet lists.

No. it can send no more than `device->efa_attr.max_sq_wr`.

the packet entry pool's size is configured as this size, so as along as we can allocate a packet entry, we should be able to call `ibv_post_send`

File: prov/efa/src/rdm/rxr_pkt_cmd.c [hunk 1]
```diff
@@ -308,7 +308,7 @@ ssize_t rxr_pkt_post_one(struct efa_rdm_ep *efa_rdm_ep, struct efa_rdm_ope *ope,
 	/* If the send succeeded, the function rxr_pkt_entry_send will increase the
 	 * counter in efa_rdm_ep that tracks number of outstanding TX ops.
 	 */
-	err = rxr_pkt_entry_send(efa_rdm_ep, pkt_entry, flags);
+	err = rxr_pkt_entry_send(efa_rdm_ep, &pkt_entry, 1, flags);
```
Comment (shijin-aws) [inline prov/efa/src/rdm/rxr_pkt_cmd.c:311]: I feel like we should rename this function to rxr_pkt_entry_sendv, to match the libfabric and MPI naming convention

**Comment by shijin-aws** [comment]
Is there a future PR to call `rxr_pkt_entry_send` with multiple pkts?

File: prov/efa/src/rdm/rxr_pkt_cmd.c (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/rdm/rxr_pkt_cmd.c:311]: sure. sounds good.

**Comment by wzamazon** [comment]
> Is there a future PR to call rxr_pkt_entry_send with multiple pkts?

Yes. I am working on it.

**Comment by wzamazon** [comment]
The next PR will get rid of the usage of FI_MORE internally.

File: prov/efa/src/rdm/rxr_pkt_cmd.c (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/rdm/rxr_pkt_cmd.c]: addressed in new revision

**Review by shijin-aws — APPROVED** [review]

File: prov/efa/src/rdm/rxr_pkt_entry.c [hunk 2]
```diff
@@ -364,82 +364,94 @@ void rxr_pkt_entry_append(struct rxr_pkt_entry *dst,
  * @brief Populate pkt_entry->ibv_send_wr with the information stored in pkt_entry,
```
Comment (wenduwan) [inline prov/efa/src/rdm/rxr_pkt_entry.c:364]: > Populate pkt_entry->ibv_send_wr with the information stored in pkt_entry

This is no longer accurate.

File: prov/efa/src/rdm/rxr_pkt_entry.c [hunk 3]
```diff
@@ -364,82 +364,94 @@ void rxr_pkt_entry_append(struct rxr_pkt_entry *dst,
...
+		send = pkt_entry->send;
+
+		/* EFA device supports a maximum of 2 iov/SGE
+		 */
+		assert(send->iov_count <= 2);
+
+		peer = efa_rdm_ep_get_peer(ep, pkt_entry->addr);
+		assert(peer);
+		if (peer->flags & EFA_RDM_PEER_IN_BACKOFF)
+			return -FI_EAGAIN;
```
Comment (wenduwan) [inline prov/efa/src/rdm/rxr_pkt_entry.c:402]: Returning error in this loop is problematic.

It reads to me that, it is possible the function leaves `ep->base_ep.xmit_more_wr_tail` in an unknown state if it returns an error, e.g. we appended 10 send_wr to it but then returns an error to the caller.

Am I missing anything? I'm trying to understand the atomicity of this function.

Comment (wzamazon) [reply prov/efa/src/rdm/rxr_pkt_entry.c:402]: You are right this should have been moved to the before the loop.

Comment (wzamazon) [reply prov/efa/src/rdm/rxr_pkt_entry.c]: addressed in new revision.

**Comment by wenduwan** [comment]
bot:aws:retest

**Pull request merged**